### ISSUE TITLE
Get data from nanoos 52nsos

### ISSUE THREAD

**Issue by ocefpaf** [ocefpaf/sscsw#6]
Collect the data from

http://data.nanoos.org/52nsos/sos/kvp?service=SOS&request=GetCapabilities&acceptVersions=1.0.0

I am trying to follow @emiliom's notebook [1], but my attempt is timing out.  Not sure what I am doing wrong there.

[1] https://www.wakari.io/sharing/bundle/emayorga/pyoos_ioos_sos_demo1
[2] http://nbviewer.ipython.org/gist/ocefpaf/5e59d4b5679ec72aa8c1

**Comment by emiliom** [comment]
I'm working on both fixing the time out, and developing code to use to ingest the nanoos 52n sos time series.

**Comment by ocefpaf** [comment]
Don't worry to much about it. I believe Kathy has what she needs.  I am opening issues just to keep track of things.

**Comment by emiliom** [comment]
Yes. I figured the NANOOS SOS results won't make it into this notebook, for Kathy's presentation. But I want to get to the bottom of these problems, and if I don't focus now, I'll forget what I've learned in the last couple of days. I already made a ton of progress in the last 24 hours!

**Comment by emiliom** [comment]
- The good news: I've found a solution to the **timeout problem**. pyoos (more specifically, owslib) is implicitly relying on the default timeout value from `requests` for GET requests. But multi-station SOS requests can be slow, and my 52North SOS server hasn't been optimized beyond having enough memory allocation. So, the request times out due to a self-imposed limit, not b/c of a server problem. The [solution](https://github.com/geopython/OWSLib/blob/master/owslib/swe/observation/sos100.py#L187) is to pass an optional timeout argument to the collect method, like this: `timeout=100`. I specifically tested that value on a request w/o a bbox filter. _I think we should tweak pyoos to set a new, longer default timeout; SOS requests can get long._ This would be an easy tweak, [probably here](https://github.com/ioos/pyoos/blob/master/pyoos/collectors/ioos/swe_sos.py#L34)
- The bad news: the **bbox filter** [implementation in pyoos IoosSweSos](https://github.com/ioos/pyoos/blob/master/pyoos/collectors/ioos/swe_sos.py#L38) doesn't seem to work with 52North SOS! I've researched this (eg, [here](https://wiki.52north.org/bin/view/SensorWeb/GetObservation#Examples_HTTP_GET)) and tried different options via direct HTTP POST and GET requests. The 52N SOS server does have a bbox query that I can successfully use via POST; but I can't find a GET alternative (pyoos/owslib relies on GET SOS requests), and what pyoos has implemented definitely fails. I'll follow up with Shane (Axiom), the IOOS 52N SOS implementer/developer. When I find a solution I'll implement it on pyoos and submit a pull request. Hopefully within a week ...

**Comment by ocefpaf** [comment]
Thanks for looking further into this!

Timeouts  greater than 30 seconds are out of question for demonstrations and interactive work. I recommend against changing the default because the user can always set a higher value when s/he knows it will take longer.

(When something takes > 30 s, in interactive work, people just re-start the process before its end thinking that the action was hanging up on them.)

PS: I got [some data](http://nbviewer.ipython.org/gist/ocefpaf/5e59d4b5679ec72aa8c1) by setting a higher timeout. I will see if I can write something later to collect all the data in a DataFrame.

**Comment by emiliom** [comment]
> Timeouts greater than 30 seconds are out of question for demonstrations and interactive work.

I agree. But in principle, a notebook (not specifically for the OOI workshop) may be viewed statically via nbviewer. Notebooks like this provide code illustrations for others to grab and implement, not just for interactive viewing.

> I recommend against changing the default because the user can always set a higher value when s/he knows it will take longer.

That's true. The problem I see is that pyoos makes it so darn difficult to use object introspection (or consult the documentation ...) to know what's going on, that it would take a user a lot of work -- like it did with me -- to figure out there's a timeout parameter and a default value that's being exceeded. 

But maybe a compromise solution is to add a docstring to pyoos `IoosSweSos.collect`, which tells the user about the existence of the timeout parameter and its default value ...

> PS: I got some data by setting a higher timeout.

Cool! And great to see your progress on that notebook. 

But beware that you'll run into funkyness if you're paying attention to the data. I think I've learned that pyoos IoosSweSos is not properly handling stations with measurements at multiple depths (eg, >1 temp sensors along a mooring anchor line); the nanoos sos returns two of those in the ooi bbox. It's also possible that the 52N SOS, or the way I push data to my instance, is not following the IOOS SOS standard properly. I'll investigate further, and plan on submitting a pull request to pyoos if I find a problem. But that's longer term, not next week!

You can also just wait until next week to let me provide some code for this (but not the multi-depth handling part) ...

**Comment by emiliom** [comment]
FYI, for your notebook: You no longer need to define the `flatten_element` function (your Cell 18), if you were using the master branch of pyoos. I added it as a utility function a year ago. You just do this:

``` python
from pyoos.utils.dataorg import flatten_element
```

That'll be available once a new pyoos release is made, and the pyoos conda package is updated.

**Comment by emiliom** [comment]
_Note for future reference._ Regarding the optional `timeout` kwarg and the occurrence of timeouts with other pyoos collectors: looks like it had [already been pointed out, in the context of the NDBC collector](https://github.com/ioos/pyoos/issues/47). More motivation to expose it in the documentation, but not just in the IoosSweSos collector docstring. Maybe also to the [skeleton wiki](https://github.com/ioos/pyoos/wiki) that's there.

**Comment by ocefpaf** [comment]
> The problem I see is that pyoos makes it so darn difficult to use object introspection (or consult the documentation ...) to know what's going on ... solution is to add a docstring...

We do need to add docstrings everywhere and start building some docs page.  Any software is as good as its docs!

I created a new issue label (documentation) and I will collect a few issue there and tackle them as soon as I understand a little bit more about pyoos.

> You no longer need to define the `flatten_element`

Thanks!

**Comment by emiliom** [comment]
@ocefpaf FYI here are two of my follow-ups regarding the 52North SOS:
- BBOX query: https://github.com/ioos/i52n-sos/issues/25
- z encoding: https://github.com/ioos/i52n-sos/issues/24

No action on your part is needed. Just keeping you in the loop.

**Comment by ocefpaf** [comment]
Thanks @emiliom!

PS: This notebook would not be possible with you help!

**Comment by emiliom** [comment]
> PS: This notebook would not be possible with you help!

Thanks!

I noticed you've closed this issue. I wanted to be clear that you don't go away thinking that [your notebook](http://nbviewer.ipython.org/gist/ocefpaf/5e59d4b5679ec72aa8c1) works totally fine with any Ioos SOS response. I've identified a number of problems (probably bugs) involving how i52n servers encode "heights" and how pyoos parses heights (from either ncsos or i52n, it doesn't matter). For simple stations that have sensors at only one height or depth (at least per variable), this is not a big problem. But for stations that measure the same variable at >1 depth (say, a mooring with two CTD sensors), it IS a big problem and important series of bugs! Regarding pyoos, I already identified and have an easy solution for one of the bugs; I've looked into the other, more important problem, but need to confirm that there really is a bug and then code a solution (I have a path forward already). I won't get to this until at least next week, probably later. Stay tuned ...

**Comment by ocefpaf** [comment]
> I noticed you've closed this issue. I wanted to be clear that you don't go away thinking that your notebook works totally fine with any Ioos SOS response

Hehe.  Not a chance!  I just wanted to close the "I asked @emiliom for help part."

I plan to open separate issues for all the points we raised here.

**Comment by ocefpaf** [comment]
Ha!  We are making these so long that only now I realized I meant to close #3 and not #6.
